package NYT.NHiveClient.NProto;

import "yt_proto/yt/core/misc/proto/guid.proto";
import "yt_proto/yt/core/tracing/proto/tracing_ext.proto";
import "yt/ytlib/hive/proto/cell_directory.proto";

////////////////////////////////////////////////////////////////////////////////

message TEncapsulatedMessage
{
    required string type = 1;
    required bytes data = 2;
    reserved 3 to 8;
    optional NYT.NTracing.NProto.TTracingExt tracing_ext = 9;
    optional int64 logical_time = 10 [default = 0];
}

////////////////////////////////////////////////////////////////////////////////

message TReqPing
{
    required NYT.NProto.TGuid src_cell_id = 1;
}

message TRspPing
{
    optional int64 last_outcoming_message_id = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqSyncCells
{
    repeated NHiveClient.NProto.TCellInfo known_cells = 2;
}

message TUnregisterCellInfo
{
    required NYT.NProto.TGuid cell_id = 1;
}

message TReconfigureCellInfo
{
    required NYT.NHiveClient.NProto.TCellDescriptor cell_descriptor = 1;
}

message TRspSyncCells
{
    repeated TUnregisterCellInfo cells_to_unregister = 4;
    repeated TReconfigureCellInfo cells_to_reconfigure = 5;
}

////////////////////////////////////////////////////////////////////////////////

message TReqPostMessages
{
    required NYT.NProto.TGuid src_endpoint_id = 1;

    required int64 first_message_id = 2;
    repeated TEncapsulatedMessage messages = 3;

    repeated TReqPostMessages avenue_subrequests = 4;
}

message TRspPostMessages
{
    optional int64 next_persistent_incoming_message_id = 1;
    required int64 next_transient_incoming_message_id = 2;

    repeated TRspPostMessages avenue_subresponses = 3;
}

////////////////////////////////////////////////////////////////////////////////

message TReqSendMessages
{
    required NYT.NProto.TGuid src_cell_id = 1;
    repeated TEncapsulatedMessage messages = 2;
}

message TRspSendMessages
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqSyncWithOthers
{
    repeated NYT.NProto.TGuid src_cell_ids = 1;
}

message TRspSyncWithOthers
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqGetConsistentState
{
    optional int64 logical_time = 1;
}

message TRspGetConsistentState
{
    required int64 logical_time = 1;
    required int64 sequence_number = 2;
    required int32 segment_id = 3;
}

////////////////////////////////////////////////////////////////////////////////
